rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,833 @@
+
+Utilizing an extensible system for expressing the schema of Vector's configuration gives us the
+runway to solve our current problems, but also the ability to handle future problems and
+requirements as they come up.
+
+## Drawbacks
+
+The primary drawback of this approach is that it limits us to things we can reasonably express
+within the limitations of the Rust language itself, and what we're comfortable with representing via
+attributes. For example, single-line and multi-line doc comments are trivially supported, but if we
+wanted to start pushing more semantically-relevant information, such as configuration snippets, etc,
+it could be technically possible but appear as a very ugly attribute usage, which could make the
+source muddy and unclear.
+
+Additionally, while derive macros are written in Rust code, which can be much easier to grok than
+declarative macros, it represents a section of code that may be harder to understand and modify than
+if we used a more brute force approach.
+
+## Prior Art
+
+Notably, and perhaps most obviously, Cue itself can be used for data validation purposes. There are
+also other projects that slot themselves into the same general domain, such as [OPA/Rego][opa_rego]
+and [CDDL][cddl].
+
+The primary problem with all of these solutions is that they're all custom languages, with far less
+cross-platform support, and a far steeper learning curve. Even though our approach generally
+concerns itself with obfuscating the schema tool itself, and focusing on making it trivially to
+generate the schema from annotated Rust types/fields, eventually the rubber must meet the road, and
+this is where these other tools would fall down for our case and become very hard to wield
+correctly.
+
+## Alternatives
+
+### Use `schemars` only
+
+While `schemars` itself has support for annotating Rust types in almost the exact same way as we've
+proposed above, it lacks a few features necessary for our use case:
+
+- no support for the serde `alias` attribute feature
+- no support for defining generic metadata for a type/field and exposing it in the schema
+- no mechanism to override constraints for a field which already defines its own constraints at the
+  type level
+
+The lack of these features make it much harder to correctly generate our schema, as we would still
+be required to do the minimum amount of work to support defining custom attributes for the missing
+features, and the work to support at least one attribute is 90% of the work to support two, or
+three, or more, custom attributes. It would also mean that developers would need to figure out
+whether or not `schemars` provided a certain behavior via its attributes, versus being able to only
+need to remember how to get to the documentation for the our proposed internal approach.
+
+### Generate the Rust configuration types from a non-Rust source of truth
+
+Another alternative is the possibility of moving the configuration source-of-truth outside of Rust
```

Comment:
(Reposting it here for others to be able to easily see)

From @zsherman:

> Are there any contenders for what this might be?

Practically speaking, the primary and non-negotiable requirement is that it would require broad, cross-platform to be useful, otherwise it would be impossible for it to provide us net positive value. The most obvious options that fit that requirement would be JSONSchema and Protocol Buffers.

The problem with Protocol Buffers, and most other message serialization definition languages, is that it does not have the concept of applying constraints, such as "this value can be between 0 and 42".  The only available constraints are the ones baked into the data types themselves i.e. a 16-bit unsigned integer can be 0 to 2^16, etc.

Another problem with both JSONSchema and Protocol Buffers is that it would be very hard to write a type generator that could utilize `serde` effectively.  While we have a fair amount of freedom on the Rust side to parse basic serialized values and expand them into much-richer Rust types, using JSONSchema or Protocol Buffers to do the same would be very hard, or potentially even impossible, because if you're using an intentionally succinct definition language... how do you signal to one specific consumer of that definition to interpret a particular construct in a certain way?  And if you're adding consumer-specific annotations like that, what benefit are you receiving by using the stripped down, maximally generic definition language in the first place?